add pt2 callbacks for backward pass

burak-turk opened the pull request:
Summary: This change adds callbacks for lazy backwards compilation.

Differential Revision: D68515699




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

burak-turk commented:
@pytorchbot label "topic: not user facing"

mlazos commented on torch/_inductor/compile_fx.py:
```diff
@@ -563,6 +563,7 @@ def compile_fx_inner(
         # the counter here because we may dropped into compile_fx directly
         # from lazy backwards compilation.
         stack.enter_context(_WaitCounter("pytorch.wait_counter.dynamo_compile").guard())
+        stack.enter_context(torch._dynamo.callback_handler.install_callbacks())
```
I mentioned in chat, we should only do this if is_backward is true, or else we will do this twice for the forward pass. This might be okay but it's better to be safe than sorry

burak-turk replied:
I created a new PR with a different approach to duplicate callbacks prevention and unit tests: https://github.com/pytorch/pytorch/pull/145732. 

Closing this one.

burak-turk committed:
add pt2 callbacks for backward pass (#145427)

Summary:

This change adds callbacks for lazy backwards compilation.

Differential Revision: D68515699
